include/bls12-381/scalar.hpp

```diff
@@ -215,21 +215,39 @@ void rsh(std::array<uint64_t, N>& out, const std::array<uint64_t, N>& in, uint64
 
 void bn_divn_low(uint64_t *c, uint64_t *d, uint64_t *a, int sa, uint64_t *b, int sb);
 
+template<size_t N, size_t M>
+void bn_divn_safe(std::array<uint64_t, N>& c, std::array<uint64_t, M>& d, const std::array<uint64_t, N>& a, const std::array<uint64_t, M>& b) {
+    
+    static_assert(N >= M, "dividend must be at least same word size as modulus");
+
+    std::array<uint64_t, N+2> modulus = {0};
+    std::array<uint64_t, N+2> dividend = {0};
+
+    memcpy(dividend.data(), a.data(), N * sizeof(uint64_t));
+    memcpy(modulus.data(), b.data(), M * sizeof(uint64_t));
+    
+#ifdef __USE_GMP
+    // GMP version is safe
+    bn_divn_low(c.data(), d.data(), dividend.data(), N, modulus.data(), M);
+#else
+    // Relic version is unsafe
+    std::array<uint64_t, N+2> quotient = {0};
+    std::array<uint64_t, N+2> remainder = {0};
+
+    bn_divn_low(quotient.data(), remainder.data(), dividend.data(), N, modulus.data(), M);
+
+    memcpy(c.data(), quotient.data(), N * sizeof(uint64_t));
+    memcpy(d.data(), remainder.data(), M * sizeof(uint64_t));
+#endif
+}
+
 template<size_t N>
 fp fp::modPrime(std::array<uint64_t, N> k)
```

Comment:
do we need a copy here?

Reply:
k is input and can be const (and marked as const in argument list).
We only output remainder and quotient here.

Reply:
so it seems that you don't need to pass it by value

Reply:
exactly, i missed that.